### PR TITLE
[Feature] Expose XGrammar bitmask backend selection

### LINKED ISSUE #37650: [Feature]: Allow user selection of structured output (xgrammar) backend for bitmask application
### Background
In vLLM, when structured output tasks are enabled (such as using a JSON schema, grammar, or regex constraint), the backend used to apply the structured output token bitmask is an important source of both correctness and performance. The function `xgr.apply_token_bitmask_inplace` from xgrammar supports different backends (e.g., "auto", "cpu", "cuda", "triton", "torch_compile", "torch_native"), but currently in vLLM the backend is hard-coded as "auto" and not exposed to the user.

### Motivation
* Users may want to explicitly choose a backend to maximize performance for their hardware or to debug incompatibility issues, e.g., forcing `cpu` or `cuda` mode if they hit issues with the triton kernel or auto selection.
* Exposing this option provides more transparency and flexibility for power users.

### Pitch
* Allow structured output backend selection for bitmask application by exposing a command-line flag, config file option, or Python API parameter (e.g., `--structured-outputs-config.bitmask_backend` or `StructuredOutputsConfig(bitmask_backend=...)`).
* This value should be plumbed so it is finally passed as the `backend` argument to `xgr.apply_token_bitmask_inplace`, impacting calls such as the following:

```python
xgr.apply_token_bitmask_inplace(logits, grammar_bitmask, indices=index_tensor, backend=user_selected_backend)
```

* The option should be documented and behave like other StructuredOutputsConfig choices. Default behavior should remain "auto" for compatibility.

### Alternatives

- Relying only on `auto` backend selection, hoping libraries do the right thing.
- Forcing users to patch the code manually if they require a specific backend.
- Exposing the option only in Python and not via the CLI/config.

### Additional context

Relevant context: The interface for xgrammar is

```python
def apply_token_bitmask_inplace(
    logits: torch.Tensor,
    bitmask: torch.Tensor,
    *,
    vocab_size: Optional[int] = None,
    indices: Optional[List[int]] = None,
    backend: Literal["auto", "cpu", "cuda", "triton", "torch_compile", "torch_native"] = "auto",
) -> None:
```

Currently, vllm/v1/structured_output/utils.py uses it as:
```python
xgr.apply_token_bitmask_inplace(logits, grammar_bitmask, indices=index_tensor)
```

But does not pipe through a configurable backend. The `StructuredOutputsConfig` class already exists for related config. 

Let me know if additional context or example code is needed.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by xyDong0223**
## Purpose

FIX #37650

This refreshes the original implementation on the current `main` branch. It adds `--structured-outputs-config.bitmask_backend` so users can choose the XGrammar implementation used to apply token bitmasks.

The default remains `auto`. On Model Runner V2, `auto` keeps vLLM's existing local Triton kernel. An explicit value routes the operation through XGrammar so users can try another implementation when diagnosing a backend-specific problem.

The platform hook is centralized in `Platform.get_xgrammar_bitmask_backend()`. Hardware platforms can override the hook when they need a different policy, without duplicating the default implementation across CPU, CUDA, ROCm, and XPU.

## Changes

- Add the typed `bitmask_backend` field to `StructuredOutputsConfig`.
- Pass the resolved value through both the legacy model runner and Model Runner V2.
- Align Model Runner V2's request-ordered masks with XGrammar's logit-indexed interface before dispatching an explicit backend.
- Document the CLI option, supported values, and its relationship to `structured_outputs_config.backend`.
- Add unit coverage for config validation, the platform resolver, and Model Runner V2 mask alignment.

## Test Plan

```bash
python -m pre_commit run --files \
  docs/features/structured_outputs.md \
  tests/v1/structured_output/test_validation.py \
  tests/v1/worker/test_gpu_structured_outputs.py \
  vllm/config/structured_outputs.py \
  vllm/platforms/interface.py \
  vllm/v1/structured_output/utils.py \
  vllm/v1/worker/gpu/model_runner.py \
  vllm/v1/worker/gpu/structured_outputs.py \
  vllm/v1/worker/gpu_model_runner.py

python -m pytest -q \
  tests/v1/structured_output/test_validation.py \
  tests/v1/worker/test_gpu_structured_outputs.py
```

## Test Result

- All selected pre-commit hooks passed, including Ruff, Markdown linting, mypy, configuration validation, and the forbidden-import check.
- `26 passed` in the targeted unit tests.
- The Model Runner V2 dispatch test uses a mocked XGrammar call on CPU. GPU hardware was not available for an end-to-end backend run.

### CODE REVIEW

File: vllm/platforms/xpu.py [hunk 1]
```diff
@@ -116,6 +116,26 @@ def get_vit_attn_backend(
         )
         return AttentionBackendEnum.FLASH_ATTN
 
+    @classmethod
+    def get_supported_bitmask_backends(cls) -> list[str]:
+        return ["auto", "cpu"]
```
Comment (jikunshang) [inline vllm/platforms/xpu.py]: please add triton for xpu. thanks.

File: vllm/platforms/interface.py [hunk 2]
```diff
@@ -277,6 +277,41 @@ def get_vit_attn_backend(
         )
         return AttentionBackendEnum.TORCH_SDPA
 
+    @classmethod
+    def get_supported_bitmask_backends(cls) -> list[str]:
+        """Return the list of supported bitmask backends on this platform."""
+        return ["auto", "cpu"]
+
+    @classmethod
+    def get_bitmask_backend(cls, backend: str = "auto") -> str:
```
Comment (jikunshang) [inline vllm/platforms/interface.py]: I believe we can keep this method in interface.py only and remove it in cpu/cuda/rocm/xpu.py

File: vllm/platforms/xpu.py (hunk 1 above)
Comment (xyDong0223) [reply vllm/platforms/xpu.py]: done

File: vllm/platforms/interface.py (hunk 2 above)
Comment (xyDong0223) [reply vllm/platforms/interface.py]: Thank you for your suggestion. After giving it careful thought, I believe it would be an excellent idea to enable developers working on mainstream platforms—specifically those utilizing CUDA and/or XPU—to see which backends XGrammar currently supports on their respective platforms, and to allow them to select their preferred backend independently.

File: vllm/platforms/cpu.py [hunk 3]
```diff
@@ -465,6 +465,26 @@ def get_device_communicator_cls(cls) -> str:
     def supports_structured_output(cls) -> bool:
         return True
 
+    @classmethod
+    def get_supported_bitmask_backends(cls) -> list[str]:
```
Comment (chaunceyjiang) [inline vllm/platforms/cpu.py]: I feel this name is not very clear.
Maybe it would be better to set the name to something related to structured output.

Comment (xyDong0223) [reply vllm/platforms/cpu.py]: Your are right,it is make sense! Might be named  `get_supported_structured_backends` or `get_supported_garmmar_backends`

Comment (xyDong0223) [reply vllm/platforms/cpu.py]: Hi, I change the name to `get_supported_structured_output_backends`, I think it would be enhance readability. Furthermore, if there are other libraries capable of generating structured output—and if they offer a choice of backend implementations—this interface could potentially be reused.

**Comment by DarkLight1337** [comment]
No one in particular, feel free to approve if it makes sense

**Comment by xyDong0223** [comment]
Superseded by #54977, which rebases the feature on current `main`, keeps the platform hook centralized, supports both model-runner paths, adds focused tests and documentation, and passes local pre-commit checks.

**Comment by xyDong0223** [comment]
Hi @chaunceyjiang @DarkLight1337  I’ve refreshed this PR on the current main branch and addressed the earlier feedback by keeping the backend resolution in the shared platform interface, wiring the option through both the legacy runner and Model Runner V2, and adding focused tests and documentation. Local pre-commit checks and the targeted test suite pass. When you have time, could you please take another look? I’d especially appreciate feedback on the platform-resolver API and the V2 mask-alignment path.

File: vllm/platforms/interface.py [hunk 4]
```diff
@@ -416,6 +417,18 @@ def get_vit_attn_backend(
         )
         return AttentionBackendEnum.TORCH_SDPA
 
+    @classmethod
+    def get_xgrammar_bitmask_backend(
```
Comment (chaunceyjiang) [inline vllm/platforms/interface.py:421]: /cc @mgoin PTAL.

**Comment by xyDong0223** [comment]
@mgoin PTAL, thanks you
